mllib/src/main/scala/org/apache/spark/ml/stat/Summarizer.scala

```diff
@@ -0,0 +1,746 @@
+
+        s"Buffer(n=$n mean=${v(mean)} m2n=${v(m2n)} m2=${v(m2)} l1=${v(l1)}" +
+          s" totalCount=$totalCount totalWeightSum=$totalWeightSum" +
+          s" totalWeightSquareSum=$totalWeightSquareSum weightSum=${v(weightSum)} nnz=${vl(nnz)}" +
+          s" max=${v(max)} min=${v(min)})"
+      }
+    }
+
+  object Buffer extends Logging {
+    // Recursive function, but the number of cases is really small.
+    def fromMetrics(requested: Seq[ComputeMetrics]): Buffer = {
+      if (requested.isEmpty) {
+        new Buffer()
+      } else {
+        val b = fromMetrics(requested.tail)
+        requested.head match {
+          case ComputeMean => b.copy(mean = Array.empty)
+          case ComputeM2n => b.copy(m2n = Array.empty)
+          case ComputeM2 => b.copy(m2 = Array.empty)
+          case ComputeL1 => b.copy(l1 = Array.empty)
+          case ComputeWeightSum => b.copy(weightSum = Array.empty)
+          case ComputeNNZ => b.copy(nnz = Array.empty)
+          case ComputeMax => b.copy(max = Array.empty)
+          case ComputeMin => b.copy(min = Array.empty)
+          case _ => b // These cases are already being computed
+        }
+      }
+    }
+
+    /**
+     * (testing only). Makes a buffer with all the metrics enabled.
+     */
+    def allMetrics(): Buffer = {
+      fromMetrics(Seq(ComputeMean, ComputeM2n, ComputeM2, ComputeL1,
+        ComputeWeightSum, ComputeNNZ, ComputeMax,
+        ComputeMin))
+    }
+
+    val bufferSchema: StructType = {
+      val fields = Seq(
+        "n" -> IntegerType,
+        "mean" -> arrayDType,
+        "m2n" -> arrayDType,
+        "m2" -> arrayDType,
+        "l1" -> arrayDType,
+        "totalCount" -> LongType,
+        "totalWeightSum" -> DoubleType,
+        "totalWeightSquareSum" -> DoubleType,
+        "weightSum" -> arrayDType,
+        "nnz" -> arrayLType,
+        "max" -> arrayDType,
+        "min" -> arrayDType
+      )
+      StructType(fields.map { case (name, t) => StructField(name, t, nullable = true)})
+    }
+
+    val numFields = bufferSchema.fields.length
+
+    def updateInPlace(buffer: Buffer, v: Vector, w: Double): Unit = {
+      val startN = buffer.n
+      if (startN == -1) {
+        // The buffer was not initialized, we initialize it with the incoming row.
+        fillBufferWithRow(buffer, v, w)
+        return
+      } else {
+        require(startN == v.size,
+          s"Trying to insert a vector of size $v into a buffer that " +
+            s"has been sized with $startN")
+      }
+      val n = buffer.n
+      assert(n > 0, n)
+      // Always update the following fields.
+      buffer.totalWeightSum += w
+      buffer.totalCount += 1
+      buffer.totalWeightSquareSum += w * w
+      // All the fields that we compute on demand:
+      // TODO: the most common case is dense vectors. In that case we should
+      // directly use BLAS instructions instead of iterating through a scala iterator.
+      v.foreachActive { (index, value) =>
```

Comment:
RDD's Summarizer doesn't have BLAS optimization actually. So this may not be the reason for the performance gap.